src/runtime/threadpool/builder.rs

```diff
@@ -163,7 +163,21 @@ impl Builder {
         let compat_reactor = compat_bg.reactor().clone();
         let compat_sender: Arc<RwLock<Option<super::CompatSpawner<tokio_02::runtime::Handle>>>> =
             Arc::new(RwLock::new(None));
-        let compat_sender2 = compat_sender.clone();
+
+        // We need a weak ref here so that when we pass this into the
+        // `on_thread_start` closure its stored as a _weak_ ref and not
+        // a strong one. This is important because tokio 0.2's runtime
+        // should shutdown when the compat Runtime has been dropped. There
+        // can be a race condition where we hold onto an extra Arc which holds
+        // a runtime handle beyond the drop. This casuses the tokio runtime to
+        // not shutdown and leak fds.
+        //
+        // Tokio's threaded_scheduler will spawn threads that each contain a arced
+        // ref to the `on_thread_start` fn. If the runtime shutsdown but there is still
+        // access to a runtime handle the mio driver will not shutdown. To avoid this we
+        // only want the `on_thread_start` to hold a weak ref and attempt to check async if
```

Comment:
what do you mean by "check async"?

Reply:
So this fn gets run at some point async in the future and we have no control over when it gets called. Because of this I want to point out that it may happen but you don't know when.